File: src/services/ranking/attributes.ts
/**
 * Hunter attributes -> Hunter Rank (PLAN.txt §6.7–6.9). Strength/Physique/
 * Endurance/Discipline each 0..100, combined by HUNTER_WEIGHTS. All models here
 * are PROVISIONAL and use healthy, non-punitive ranges — no medical claims.
 */
import { clampScore, scoreToRank, type Rank } from '@/constants/ranks';
import { HUNTER_WEIGHTS, type Sex } from '@/services/ranking/config';
import { weightedRenormalized } from '@/services/ranking/gateScore';
import { interpolate, type Anchor } from '@/services/ranking/interp';

export interface HunterAttributes {
  strength: number | null;
  physique: number | null;
  endurance: number | null;
  discipline: number | null;
}

/** Strength attribute: average of the user's ranked exercise scores. */
export function strengthScore(exerciseScores: number[]): number | null {
  if (exerciseScores.length === 0) return null;
  const sum = exerciseScores.reduce((a, b) => a + clampScore(b), 0);
  return clampScore(sum / exerciseScores.length);
}

// PROVISIONAL healthy body-fat curves (bodyFat% -> score). Peak in a healthy
// band; not rewarding for ever-lower body fat.
const BODYFAT_MALE: readonly Anchor[] = [
  { x: 4, y: 60 },
  { x: 8, y: 85 },
  { x: 12, y: 100 },
  { x: 18, y: 90 },
  { x: 25, y: 65 },
  { x: 32, y: 40 },
  { x: 40, y: 20 },
];
const BODYFAT_FEMALE: readonly Anchor[] = [
  { x: 12, y: 60 },
  { x: 16, y: 85 },
  { x: 22, y: 100 },
  { x: 28, y: 90 },
  { x: 34, y: 65 },
  { x: 40, y: 40 },
  { x: 48, y: 20 },
];
// PROVISIONAL muscle development: skeletal-muscle mass as % of bodyweight.
const MUSCLE_DEV: readonly Anchor[] = [
  { x: 30, y: 30 },
  { x: 38, y: 60 },
  { x: 44, y: 90 },
  { x: 50, y: 100 },
];

export interface PhysiqueInput {
  bodyFatPercent: number | null;
  skeletalMuscleMassKg: number | null;
  weightKg: number | null;
  sex: Sex;
}

/** Physique attribute from body composition. Null when there is no assessment. */
export function physiqueScore(a: PhysiqueInput | null): number | null {
  if (!a) return null;
  const parts: { value: number | null; weight: number }[] = [];

  if (a.bodyFatPercent != null) {
    const curve = a.sex === 'female' ? BODYFAT_FEMALE : BODYFAT_MALE;
    parts.push({ value: clampScore(interpolate(curve, a.bodyFatPercent)), weight: 0.5 });
  }
  if (a.skeletalMuscleMassKg != null && a.weightKg != null && a.weightKg > 0) {
    const smmPct = (a.skeletalMuscleMassKg / a.weightKg) * 100;
    const scale = a.sex === 'female' ? 1.1 : 1.0; // provisional: women carry less SMM%
    parts.push({ value: clampScore(interpolate(MUSCLE_DEV, smmPct * scale)), weight: 0.5 });
  }
  if (parts.length === 0) return null;
  return weightedRenormalized(parts);
}

// PROVISIONAL endurance from weekly training minutes (guideline ~150 min).
const ENDURANCE_MINUTES: readonly Anchor[] = [
  { x: 0, y: 0 },
  { x: 75, y: 45 },
  { x: 150, y: 70 },
  { x: 300, y: 100 },
];

/** Endurance attribute from recent training volume. Null when no data. */
export function enduranceScore(weeklyTrainingMinutes: number | null): number | null {
  if (weeklyTrainingMinutes == null) return null;
  return clampScore(interpolate(ENDURANCE_MINUTES, weeklyTrainingMinutes));
}

/** Combined Hunter score, renormalized over available attributes. */
export function hunterScore(attrs: HunterAttributes): number {
  return weightedRenormalized([
    { value: attrs.strength, weight: HUNTER_WEIGHTS.strength },
    { value: attrs.physique, weight: HUNTER_WEIGHTS.physique },
    { value: attrs.endurance, weight: HUNTER_WEIGHTS.endurance },
    { value: attrs.discipline, weight: HUNTER_WEIGHTS.discipline },
  ]);
}

export function hunterRank(score: number): Rank {
  return scoreToRank(score);
}

export type AttributeName = keyof HunterAttributes;

/** The lowest available attribute — the one most limiting the next rank. */
export function limitingAttribute(attrs: HunterAttributes): AttributeName | null {
  let name: AttributeName | null = null;
  let lowest = Infinity;
  (Object.keys(attrs) as AttributeName[]).forEach((k) => {
    const v = attrs[k];
    if (v != null && v < lowest) {
      lowest = v;
      name = k;
    }
  });
  return name;
}

File: android-native/app/src/main/java/com/fitnessrpg/app/domain/ranking/Attributes.kt
package com.fitnessrpg.app.domain.ranking

import com.fitnessrpg.app.domain.rank.Rank
import com.fitnessrpg.app.domain.rank.clampScore
import com.fitnessrpg.app.domain.rank.scoreToRank

/**
 * Hunter attributes -> Hunter Rank (PLAN.txt §6.7-6.9). Strength/Physique/
 * Endurance/Discipline each 0..100, combined by [HunterWeights]. All models here
 * are PROVISIONAL and use healthy, non-punitive ranges — no medical claims.
 */
data class HunterAttributes(
    val strength: Double?,
    val physique: Double?,
    val endurance: Double?,
    val discipline: Double?,
)

/** Strength attribute: average of the user's ranked exercise scores. */
fun strengthScore(exerciseScores: List<Double>): Double? {
    if (exerciseScores.isEmpty()) return null
    val sum = exerciseScores.sumOf { clampScore(it) }
    return clampScore(sum / exerciseScores.size)
}

// PROVISIONAL healthy body-fat curves (bodyFat% -> score). Peak in a healthy
// band; not rewarding for ever-lower body fat.
private val BODYFAT_MALE: List<Anchor> = listOf(
    Anchor(4.0, 60.0), Anchor(8.0, 85.0), Anchor(12.0, 100.0), Anchor(18.0, 90.0),
    Anchor(25.0, 65.0), Anchor(32.0, 40.0), Anchor(40.0, 20.0),
)
private val BODYFAT_FEMALE: List<Anchor> = listOf(
    Anchor(12.0, 60.0), Anchor(16.0, 85.0), Anchor(22.0, 100.0), Anchor(28.0, 90.0),
    Anchor(34.0, 65.0), Anchor(40.0, 40.0), Anchor(48.0, 20.0),
)

// PROVISIONAL muscle development: skeletal-muscle mass as % of bodyweight.
private val MUSCLE_DEV: List<Anchor> = listOf(
    Anchor(30.0, 30.0), Anchor(38.0, 60.0), Anchor(44.0, 90.0), Anchor(50.0, 100.0),
)

data class PhysiqueInput(
    val bodyFatPercent: Double?,
    val skeletalMuscleMassKg: Double?,
    val weightKg: Double?,
    val sex: String?,
)

/** Physique attribute from body composition. Null when there is no assessment. */
fun physiqueScore(a: PhysiqueInput?): Double? {
    if (a == null) return null
    val parts = mutableListOf<WeightedComponent>()

    if (a.bodyFatPercent != null) {
        val curve = if (a.sex == "female") BODYFAT_FEMALE else BODYFAT_MALE
        parts.add(WeightedComponent(clampScore(interpolate(curve, a.bodyFatPercent)), 0.5))
    }
    val smm = a.skeletalMuscleMassKg
    val weight = a.weightKg
    if (smm != null && weight != null && weight > 0.0) {
        val smmPct = (smm / weight) * 100.0
        val scale = if (a.sex == "female") 1.1 else 1.0 // provisional: women carry less SMM%
        parts.add(WeightedComponent(clampScore(interpolate(MUSCLE_DEV, smmPct * scale)), 0.5))
    }
    if (parts.isEmpty()) return null
    return weightedRenormalized(parts)
}

// PROVISIONAL endurance from weekly training minutes (guideline ~150 min).
private val ENDURANCE_MINUTES: List<Anchor> = listOf(
    Anchor(0.0, 0.0), Anchor(75.0, 45.0), Anchor(150.0, 70.0), Anchor(300.0, 100.0),
)

/** Endurance attribute from recent training volume. Null when no data. */
fun enduranceScore(weeklyTrainingMinutes: Double?): Double? {
    if (weeklyTrainingMinutes == null) return null
    return clampScore(interpolate(ENDURANCE_MINUTES, weeklyTrainingMinutes))
}

/** Combined Hunter score, renormalized over available attributes. */
fun hunterScore(attrs: HunterAttributes): Double = weightedRenormalized(
    listOf(
        WeightedComponent(attrs.strength, HunterWeights.STRENGTH),
        WeightedComponent(attrs.physique, HunterWeights.PHYSIQUE),
        WeightedComponent(attrs.endurance, HunterWeights.ENDURANCE),
        WeightedComponent(attrs.discipline, HunterWeights.DISCIPLINE),
    ),
)

fun hunterRank(score: Double): Rank = scoreToRank(score)

enum class AttributeName { STRENGTH, PHYSIQUE, ENDURANCE, DISCIPLINE }

/** The lowest available attribute — the one most limiting the next rank. */
fun limitingAttribute(attrs: HunterAttributes): AttributeName? {
    var name: AttributeName? = null
    var lowest = Double.POSITIVE_INFINITY
    // Iteration order matches the attribute declaration order; strict `<` keeps
    // the earlier attribute on ties.
    val pairs = listOf(
        AttributeName.STRENGTH to attrs.strength,
        AttributeName.PHYSIQUE to attrs.physique,
        AttributeName.ENDURANCE to attrs.endurance,
        AttributeName.DISCIPLINE to attrs.discipline,
    )
    for ((key, value) in pairs) {
        if (value != null && value < lowest) {
            lowest = value
            name = key
        }
    }
    return name
}
